src/mds/CDir.cc

```diff
@@ -2160,10 +2161,181 @@ class C_IO_Dir_Committed : public CDirIOContext {
+  vector<dentry_key_t> to_remove;
+  vector<CDir::dentry_commit_item> to_set;
+  mempool::mds_co::compact_set<mempool::mds_co::string> stale_items;
+};
+
+// This is not locked by mds_lock
+void CDir::_omap_commit_ops(int r, int op_prio, version_t version, bool _new, bufferlist &dfts,
+                            vector<dentry_key_t>& to_remove, vector<dentry_commit_item> &to_set,
+			    mempool::mds_co::compact_set<mempool::mds_co::string> &stales)
+{
+  dout(10) << __func__ << dendl;
+
+  if (r < 0) {
+    mdcache->mds->handle_write_error_with_lock(r);
+    return;
+  }
+
+  C_GatherBuilder gather(g_ceph_context,
+                         new C_OnFinisher(new C_IO_Dir_Committed(this, version),
+			 mdcache->mds->finisher));
+
+  SnapContext snapc;
+  object_t oid = get_ondisk_object();
+  object_locator_t oloc(mdcache->mds->mdsmap->get_metadata_pool());
+
+  map<string, bufferlist> _set;
+  set<string> _rm;
+
+  unsigned max_write_size = mdcache->max_dir_commit_size;
+  unsigned write_size = 0;
+
+  char *dftstr = dfts.c_str();
+
+  auto commit_one = [&](bool header=false) {
+    ObjectOperation op;
+
+    // don't create new dirfrag blindly
+    if (!_new)
+      op.stat(nullptr, nullptr, nullptr);
+
+    /*
+     * save the header at the last moment.. If we were to send it off before
+     * other updates, but die before sending them all, we'd think that the
+     * on-disk state was fully committed even though it wasn't! However, since
+     * the messages are strictly ordered between the MDS and the OSD, and
+     * since messages to a given PG are strictly ordered, if we simply send
+     * the message containing the header off last, we cannot get our header
+     * into an incorrect state.
+     */
+    if (header) {
+      bufferlist header;
+      encode(*fnode, header);
+      op.omap_set_header(header);
+    }
+
+    op.priority = op_prio;
+    if (!_set.empty())
+      op.omap_set(_set);
+    if (!_rm.empty())
+      op.omap_rm_keys(_rm);
+    mdcache->mds->objecter->mutate(oid, oloc, op, snapc,
+                                   ceph::real_clock::now(),
+                                   0, gather.new_sub());
+    write_size = 0;
+    _set.clear();
+    _rm.clear();
+  };
+
+  for (auto &key : stales) {
+    write_size += key.length();
+    _rm.emplace(key);
+
+    if (write_size >= max_write_size)
+      commit_one();
+  }
+
+  for (auto &k : to_remove) {
+    string key;
+    k.encode(key);
+    write_size += key.length();
+    _rm.emplace(std::move(key));
+
+    if (write_size >= max_write_size)
+      commit_one();
+  }
+
+  uint64_t off = 0;
+  bufferlist bl;
+  for (auto &item : to_set) {
+    string key;
+    item.key.encode(key);
+
+    encode(item.first, bl);
+    if (item.is_remote) {
+      bl.append('L');         // remote link
+      encode(item.ino, bl);
+      encode(item.d_type, bl);
+    } else {
+      bl.append('I');         // inode
+      {
+        std::lock_guard l(mdcache->mds->mds_lock);
```

Comment:
@ukernel @batrick 

For this, we need the `mds_lock` here. Because if the `map` is changed during the `encode()` is traversing it, we may hit the crash issue. And also the same with the following `xattrs` and `old_inodes`. I just lock them separately. 

Or should we just copy them out first and then encode them out of the `mds_lock` ?

Reply:
no, mds_lock is not needed here

Reply:
Removed it now. Thanks.